### PR TITLE
LocalPreciseDateTime improvements

### PR DESCRIPTION
**Pull request by themattchan**
- Add `diff` function for LocalPreciseDateTime. ~This is a bit of a hack which can be cleaned up after #18 is merged.~

- Fix `fromRFC3339String` for LocalPreciseDateTime to not normalize the `DateTime` component if a locale is specified. This is done by first removing the locale during parsing.

- Guard against invalid locales in the `RFC3339String.toLocale` parser.

### CODE REVIEW

**Comment by paulyoung** [comment]
Left some feedback on #18 so will wait on reviewing this.

**Comment by paulyoung** [comment]
#18 has been merged but I'm going to wait for this PR before making a v4 release, since this also includes breaking changes.

**Comment by paulyoung** [comment]
@themattchan is this ready for another review?

**Comment by themattchan** [comment]
@paulyoung yes

File: src/Data/RFC3339String.purs [hunk 1]
```diff
@@ -51,15 +53,26 @@ fromDateTime :: DateTime -> RFC3339String
...
+      guard $ zero <= hrs' && hrs' <= 24.0
+      guard $ if hrs' == 24.0 then mins' == zero else zero <= mins' && mins' <= 59.0
+      let offset = convertDuration (Hours hrs') + Minutes mins'
+      pure $ (if sign == "-" then negate else id) offset
+
+-- | Normalise the locale to GMT.
+normLocale :: RFC3339String -> RFC3339String
+normLocale (RFC3339String s) =
+  let re = RE.unsafeRegex "([-|\\+])(\\d\\d):?(\\d\\d)$" RE.noFlags
+  in RFC3339String (RE.replace re "Z" s)
```
Comment (paulyoung) [inline src/Data/RFC3339String.purs]: I'm a little bit concerned that this function could be confused with a function that also adjusts the date/time.

Perhaps there's a better name for it?

File: src/Data/PreciseDateTime/Locale.purs [hunk 2]
```diff
@@ -2,27 +2,38 @@ module Data.PreciseDateTime.Locale where
...
 type LocalPreciseDateTime = LocalValue PreciseDateTime
 
 adjust :: PreciseDuration -> LocalPreciseDateTime -> Maybe LocalPreciseDateTime
 adjust = traverse <<< PDT.adjust
 
+diff :: LocalPreciseDateTime -> LocalPreciseDateTime -> PreciseDuration
+diff (LocalValue (Locale _ m1) pdt1) (LocalValue (Locale _ m2) pdt2) =
+  let offsetDiff = PD.toDecimalLossy (PD.toNanoseconds (PD.minutes (Decimal.fromNumber (unwrap (m1 - m2)))))
+      dtDiff = PD.toDecimalLossy (PD.toNanoseconds (PDT.diff pdt1 pdt2))
+  in PD.unsafeNanoseconds (offsetDiff + dtDiff)
```
Comment (paulyoung) [inline src/Data/PreciseDateTime/Locale.purs]: Could we add tests for this?

File: src/Data/RFC3339String.purs (hunk 1 above)
Comment (themattchan) [reply src/Data/RFC3339String.purs]: How about `removeLocale`?

Comment (paulyoung) [reply src/Data/RFC3339String.purs]: In terms of the RFC3339 standard it's replacing the offset

File: src/Data/PreciseDateTime/Locale.purs (hunk 2 above)
Comment (themattchan) [reply src/Data/PreciseDateTime/Locale.purs]: Added tests.

File: src/Data/RFC3339String.purs (hunk 1 above)
Comment (themattchan) [reply src/Data/RFC3339String.purs]: Renamed to `setLocaleToZ`

**Pull request merged**